Improve auth prompt instructions

erikadoyle opened the pull request:
. . . thanks to suggestions from @AE-MS and @iambmelt.

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/419a9604627737d296b2c704ea38a381dba5c3fd/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 87 | 23 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/516bb555-4efd-43b9-b83e-1d6db4cb7c96) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/4eb412bf70a4d2f459a62b75a28f3f52b6492f24/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 85 | 25 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/ff6b0609-3f67-4876-8a0f-774408a1b015) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/1aacbcba6d0bf4a0e30eb11af040b63697a4178a/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 87 | 23 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/055ca7ce-5803-44a1-88f5-77965b50c00e) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

AE-MS commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -213,25 +213,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
 Tenant admin consent: A simple way of [consenting on behalf of an organization as a tenant admin](/azure/active-directory/manage-apps/consent-and-permissions-overview#admin-consent) is by getting [consent from admin](/azure/active-directory/manage-apps/grant-admin-consent).
 
-You can ask for consent using the Auth API. Another approach for getting Graph scopes is to present a consent dialog using our existing [third party OAuth provider authentication approach](~/tabs/how-to/authentication/auth-tab-aad.md#navigate-to-the-authorization-page-from-your-pop-up-page). This approach involves popping up an Azure AD consent dialog box.
+Another approach for getting Graph scopes is to present a consent dialog using our existing [third party OAuth provider authentication approach](~/tabs/how-to/authentication/auth-tab-aad.md#navigate-to-the-authorization-page-from-your-pop-up-page). This approach involves popping up an Azure AD consent dialog box.
```
You obviously know better than me but do we use phrases like "popping up" in documentation this? It strikes me as informal, as opposed to "showing an Azure AD consent dialog box." But again what do I know.

erikadoyle replied:
Good catch!

AE-MS commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -213,25 +213,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
+> [!TIP]
+> The [Teams Personal Tab SSO Authentication](https://github.com/OfficeDev/Microsoft-Teams-Samples/blob/main/samples/tab-personal-sso-quickstart/js/src/components/Tab.js#L64-L101) sample provides code demonstrating following steps.
 
 1. The token retrieved using `getAuthToken()` must be exchanged on the server-side using Azure AD [on-behalf-of flow](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) to get access to those other Graph APIs. Ensure you use the v2 Graph endpoint for this exchange.
```
Should this say "v2 **AAD** endpoint for this exchange?"

erikadoyle replied:
I'm not sure! The [linked doc](https://review.learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow?branch=main&branchFallbackFrom=pr-en-us-8868#success-response-example) talks about v1- and v2-formatted tokens for Microsoft Graph. I'm not able to tell if we're talking about v2 of AAD, Graph, or OAuth. 

@iambmelt could you help interpret this phrase: *Ensure you use the v2 Graph endpoint for this exchange* Thanks!

iambmelt replied:
Confirming comment from @AE-MS above - this should be the AAD v2 endpoint

AE-MS commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -213,25 +213,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
-2. If the exchange fails, Azure AD returns an invalid grant exception. It usually responds with one of the two error messages, `invalid_grant` or `interaction_required`.
-3. When the exchange fails, you must ask for consent. Use the user interface (UI) to ask the app user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog using [Silent authentication](~/concepts/authentication/auth-silent-aad.md).
-4. When asking for more consent from Azure AD, you must include `prompt=consent` in your [query-string-parameter](~/tabs/how-to/authentication/auth-silent-aad.md#get-the-user-context) to Azure AD, otherwise Azure AD wouldn't ask for other scopes.
+2. Often the first time you try to execute this token exchange for a user, Azure AD will refuse to exchange tokens because the user has not consented to giving your app permission to their data for this purpose. In these cases, your exchange will fail with either the `invalid_grant` or `interaction_required` error.
```
Do we know why/when it will be one error vs the other? (I would @ mention Brian for this but for some reason he is now showing up in the dropdown for mentions... @iambmelt)

AE-MS replied:
Oh maybe it worked after all

iambmelt replied:
The semantics for `interaction_required` are specified by [OIDC](https://openid.net/specs/openid-connect-basic-1_0.html) and is AFAIK served from the `/authorization` endpoint.

`invalid_grant` is a broader error message (sometimes remediable through user interaction) that can occur at the `/token` endpoint (e.g., `auth_code` was invalid, malformed refresh token, bad request). I'm working off memory here and can dig deeper if you're interested

AE-MS replied:
In an OBO scenario should apps expect to possibly receive either of these errors? The main thing I want to ensure the docs are clear on is what developers should expect to have to handle.

Note: in the [OBO docs](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) it tells you to hit the [/token endpoint](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#middle-tier-access-token-request) but in the [error response example](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#error-response-example) on that same page, the example error given is `interaction_required` (even though the URL it said to use was `/token` and not `/authorization`).

iambmelt replied:
Ahhh there's some interesting behavior here, Conditional Access will solicit `interaction_required` from the token endpoint if a step-up is required, as these are user remediable. BUT, for reasons unapparent to me, needing consent is an `invalid_grant`, despite the error message telling you to go interactive.

```json
{
	"error": "invalid_grant",
	"error_description": "AADSTS65001: The user or administrator has not consented to use the application with ID 'XXXXX' named 'XXXXX'. Send an interactive authorization request for this user and resource.",
	"error_codes": [65001],
	"timestamp": "2023-06-16 15:51:31Z",
	"suberror": "consent_required"
}
```

Apps can absolutely receive both of these errors - the underlying reasons for them will be different. Some examples....

- `interaction_required` - MFA or device enrollment required. 
- `invalid_grant` - consent required or auth_code, assertion, or RT is expired, revoked, malformed, or absent

erikadoyle replied:
Thanks, I'll add those examples to the explanation! Does "RT" stand for refresh token?

AE-MS replied:
I didn't write Brian's comment (obviously) but I believe that RT does indeed stand for refresh token.

AE-MS commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -213,25 +213,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
-3. When the exchange fails, you must ask for consent. Use the user interface (UI) to ask the app user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog using [Silent authentication](~/concepts/authentication/auth-silent-aad.md).
-4. When asking for more consent from Azure AD, you must include `prompt=consent` in your [query-string-parameter](~/tabs/how-to/authentication/auth-silent-aad.md#get-the-user-context) to Azure AD, otherwise Azure AD wouldn't ask for other scopes.
+2. Often the first time you try to execute this token exchange for a user, Azure AD will refuse to exchange tokens because the user has not consented to giving your app permission to their data for this purpose. In these cases, your exchange will fail with either the `invalid_grant` or `interaction_required` error.
+3. If the exchange fails for one of these two reasons, you must prompt the user for consent. Since user interaction can only happen from the client, your server will need to return an indication to your client app that consent is required. You can then use the user interface (UI) to ask the app user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog usi ...
```
Not sure if this is just because it's in review but following this silent authentication link takes me to [here](https://review.learn.microsoft.com/en-us/microsoftteams/platform/concepts/authentication/authentication?branch=main) which doesn't even contain the word "silent." 😅

erikadoyle replied:
I found the PR to remove and redirect the [AAD silent authentication](https://github.com/MicrosoftDocs/msteams-docs/pull/7408/files) article for Teams tab apps. There's no further explanation in the PR, so I'm not sure of the reason. @AE-MS, @iambmelt would you have any ideas?

erikadoyle replied:
Resolving this as I substituted a link to Teams article that [describes the consent dialog](https://learn.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/authentication/tab-sso-code#consent-dialog-for-getting-access-token) and in the next step linked to AAD article defining [consent dialog query string parameters](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow#send-the-sign-in-request).

AE-MS commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -213,25 +213,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
-3. When the exchange fails, you must ask for consent. Use the user interface (UI) to ask the app user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog using [Silent authentication](~/concepts/authentication/auth-silent-aad.md).
-4. When asking for more consent from Azure AD, you must include `prompt=consent` in your [query-string-parameter](~/tabs/how-to/authentication/auth-silent-aad.md#get-the-user-context) to Azure AD, otherwise Azure AD wouldn't ask for other scopes.
+2. Often the first time you try to execute this token exchange for a user, Azure AD will refuse to exchange tokens because the user has not consented to giving your app permission to their data for this purpose. In these cases, your exchange will fail with either the `invalid_grant` or `interaction_required` error.
+3. If the exchange fails for one of these two reasons, you must prompt the user for consent. Since user interaction can only happen from the client, your server will need to return an indication to your client app that consent is required. You can then use the user interface (UI) to ask the app user to grant other consent. This UI must include a button that triggers an Azure AD consent dialog usi ...
```
Is this correct? I am, in no sense of the word, an expert here -- but conceptually the idea of "silent" authentication triggering a dialog feels like an oxymoron?

erikadoyle replied:
Haha, fair point. I attempted to rework this and point to a current article (and screenshot): https://learn.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/authentication/tab-sso-code#consent-dialog-for-getting-access-token

AE-MS replied:
Hmmm.. I looked at the content on that page and I don't think it's helpful in this case. In this situation, I believe the developer needs to call [authentication.authenticate](https://learn.microsoft.com/en-us/javascript/api/@microsoft/teams-js/authentication?view=msteams-client-js-latest#@microsoft-teams-js-authentication-authenticate) and provide appropriate parameters to trigger the consent dialog -- notably crafting an AAD URL with the `prompt=consent` parameter on it. Not sure if there is an existing place that explains how to do this properly.

erikadoyle replied:
I added this reference from AAD docs: https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow#send-the-sign-in-request

AE-MS commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -213,25 +213,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
 Tenant admin consent: A simple way of [consenting on behalf of an organization as a tenant admin](/azure/active-directory/manage-apps/consent-and-permissions-overview#admin-consent) is by getting [consent from admin](/azure/active-directory/manage-apps/grant-admin-consent).
 
-You can ask for consent using the Auth API. Another approach for getting Graph scopes is to present a consent dialog using our existing [third party OAuth provider authentication approach](~/tabs/how-to/authentication/auth-tab-aad.md#navigate-to-the-authorization-page-from-your-pop-up-page). This approach involves popping up an Azure AD consent dialog box.
+Another approach for getting Graph scopes is to present a consent dialog using our existing [third party OAuth provider authentication approach](~/tabs/how-to/authentication/auth-tab-aad.md#navigate-to-the-authorization-page-from-your-pop-up-page). This approach involves popping up an Azure AD consent dialog box.
```
I'm actually a little confused on what this "use our existing third party...approach" means here? Is this saying "If you don't want to use OBO, then you can just directly show UI to sign the user into AAD and get a token that lets you directly access the graph?"

If that is true, is there any guidance we could give developers about why they would prefer that approach over OBO? Do we have a preference on which is a better experience (e.g., better user experience, covers more auth cases, etc.)?

erikadoyle replied:
Hmm, good questions. @iambmelt would you have any suggestions?

erikadoyle replied:
I don't understand the intent behind this (existing) statement either, and I don't think it makes sense in the current article. Its purpose seems to introduce the steps of asking the user for additional consent. I'll take a stab at reworking this--

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/80e83becb4c03af324c8577ae7fcc6a3d124f99f/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 87 | 23 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/68767bee-e477-4678-80a0-b46e2200c189) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/2bcfd9cae30033e4267c406fb152396a306a85ae/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 86 | 25 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/19dcd951-a5af-4f3c-ab89-65280003ee7b) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/54de8a05aebf6ed4da6e286a80a0d60b74e7499e/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 88 | 28 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/eeadbd72-0fd5-422c-b729-99a9b7304813) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

v-ypalikila commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -214,25 +214,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
 > - Don’t use `notifySuccess` result to return the token information to the parent page. Use `localStorage` to save the token and pass the item key via `notifySuccess`.
 
-## Known limitations
+## Obtaining consent
```
Would it be better to write Obtain consent?

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/0650f57d2437482a471ccb7819442081de908211/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 88 | 28 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/2e62db88-1245-461f-ac0a-cfb040fb8147) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/ef012c7ed4300681fd24b81764e0626becd825f1/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 88 | 28 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/0ec62845-c188-4dfd-aa29-ceb3094637d6) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/d6f169ddfd4a7badf5253bded1ee9700e47c4e2f/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 88 | 28 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/0cc79dff-1ed3-4751-bd99-27b7b4162a9f) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/1e4b47d732dcf48acab98600b232c9f9669224e1/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 88 | 27 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/c29a4ed4-aea4-456f-a4a3-0ee8aa46627f) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/d5504f20f9970b7eb0401cf780d8eed8c4edea56/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 88 | 27 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/d87614a0-e401-4d36-977a-4be14bac3ae6) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

erikadoyle commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -214,25 +214,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
+> The [Teams Personal Tab SSO Authentication](https://github.com/OfficeDev/Microsoft-Teams-Samples/blob/main/samples/tab-personal-sso-quickstart/js/src/components/Tab.js#L64-L101) sample provides code demonstrating following steps.
+
+1. The token retrieved using `getAuthToken()` must be exchanged on the server-side using Azure AD [on-behalf-of flow](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) to get access to those other Graph APIs. Ensure you use the Azure AD v2 endpoint for this exchange.
+2. Often the first time you try to execute this token exchange for a user, Azure AD will refuse to exchange tokens because the user has not consented to giving your app permission to their data for this purpose. In these cases, your exchange will fail with either the `invalid_grant` or `interaction_required` error.  Examples of *invalid_grant* errors include when consent is required or *auth_code ...
```
@v-ypalikila I've incorporated your rewrite of the sentence. Thanks!

Regarding the term "malformed", I'm not seeing guidance in either Microsoft Cloud Style Guide or general Microsoft Writing Guide against using that word, and I think it's usage is pretty standard and well understood in this domain. In which guidelines are you seeing it discouraged?

erikadoyle commented on msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md:
```diff
@@ -214,25 +214,30 @@ If you need to access Microsoft Graph data, configure your server-side code to:
+    - Ensure that `{scopes}` includes all the scopes you're prompting the user for, for example, `Mail.Read` or `User.Read`.
+
+    To handle incremental consent for tab app, see [incremental and dynamic user consent](/azure/active-directory/develop/v2-permissions-and-consent).
+5. After the app user has granted more permissions, retry the OBO flow to get access to these other APIs.
```
How about this: 

After the app user has granted more permissions, retry the OBO flow to get access to additional Graph APIs.

v-ypalikila replied:
Yes Erika, we can re-write as suggested.

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/b3de3594994bc78cca9b00cbe7e52a7ab2f52ec5/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 89 | 29 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/17fe3e9b-14f4-493a-bb41-d23b1d7ab74f) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/b15f7ec2b833e8f070259b5f87a5528f0f305cb1/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 89 | 29 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/997b7d74-aa67-4ff7-8ab9-d4b46be6d004) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

erikadoyle commented:
Apologies for the delay--I've addressed all remaining feedback. @v-ypalikila @iambmelt @AE-MS would you mind taking another look? Thanks for all the helpful insights and suggestions!

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/ebc3d1961172ae460941a5bac9c3a5009dbc123b/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 89 | 28 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/371024cb-9367-4f9b-9bcb-b9ab6fb2b575) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

acrolinxatmsft1 commented:
## Acrolinx Scorecards 
 
**We currently enforce a minimum score of 80.**
 
Click the scorecard links for each article to review the Acrolinx feedback on grammar, spelling, punctuation, writing style, and terminology:
| Article | Score | Issues | Scorecard | Processed |
| ------- | ----- | ------ | --------- | --------- |
| [msteams-platform/tabs/how-to/authentication/tab-sso-graph-api.md](https://github.com/MicrosoftDocs/msteams-docs/blob/00b30eed37a33384eebe6fd37e8a00d725f11b7d/msteams-platform%2Ftabs%2Fhow-to%2Fauthentication%2Ftab-sso-graph-api.md) | 92 | 22 | [link](https://microsoft-ce-csi.acrolinx.cloud/api/v1/checking/scorecards/738d8d97-ec10-42ce-a8d8-c546847b17c9) | [:white_check_mark:](# "Successfully processed") |

**More info about Acrolinx**
 
Use the Acrolinx extension, or sidebar, in Visual Studio Code to check spelling, grammar, style, tone, clarity, and key terminology when you're creating or updating content. For more information, see [Use the Visual Studio Code extension to run Acrolinx locally](https://review.docs.microsoft.com/office-authoring-guide/acrolinx-vscode).

v-ypalikila reviewed (approved):
Approved.
